### ISSUE TITLE
Problems with PDF Template Code in Donrec Profiles since CiviCRM 5.72.0

### ISSUE THREAD

**Issue by romymy** [systopia/de.systopia.donrec#196]
Unfortunately, I have discovered a problem that affects the processing of the template code in Donrec profiles. 

Previously, the template code could contain comments or a style tag or other structural HTML tags such as body or aside. Since CiviCRM version 5.72.0, this code is no longer displayed in the Donrec profile (but is still present in the donrec_profiles database table for an existing template and can be used for creating donation receipts). Only table, p, div, h tags or inline CSS remain, as well as tokens and Smarty variables.

However, as soon as the template is saved, the code is also replaced in the database and is hardly usable because, for example, the generation of the watermark is linked to the body tag.

This is how the first lines of the standard template should look like:
![Screenshot 2024-06-06 at 20-18-13 New Donation Receipts profile — WordPress](https://github.com/systopia/de.systopia.donrec/assets/88038958/2bd7c063-61d1-456c-8eb9-4ac9b8b2d4fb)

How they look like now:
![Screenshot 2024-06-06 at 20-19-10 Edit Donation Receipts profile — WordPress](https://github.com/systopia/de.systopia.donrec/assets/88038958/211b0e0d-9d54-4f58-9e5e-2348aa0ac0c0)

I have examined the following systems:
CiviCRM 5.69.5, donrec 2.2.7 → no problem
CiviCRM 5.71.2, donrec 2.2.8 → no problem
CiviCRM 5.72.0 (no system available at the moment)
CiviCRM 5.72.1, donrec 2.2.8 → problem with template code
CiviCRM 5.72.3, donrec 2.2.5 → problem with template code
CiviCRM 5.73.4, donrec 2.2.8 → problem with template code

From this I conclude that something has changed in the CiviCRM core with version 5.72.0, which in turn has led to the behavior of the template code field in the Donrec profile changing. Unfortunately I have not been able to identify the cause yet, but I wanted to at least share my observation. It currently means that neither new profiles can be created nor existing ones edited if CiviCRM higher than 5.72.0 is installed.

**Comment by bjendres** [comment]
Thanks for reporting @romymy. We'll have a look and see if we can reproduce this.

**Comment by alietz** [comment]
We just tested this with Civi 5.74.3. As soon as you try to edit a template, everything in the template before the <body> tag gets cut off. This is really dangerous because it's quite easy to destroy your donation templates this way!

To export your current templates, get them from the database if need be:
```
select template from donrec_profile;
```

**Comment by jofranz** [comment]
_systopia-reference: 25452_

**Comment by eileenmcnaughton** [comment]
It maybe that setPurifiedDefaults is doing this https://github.com/civicrm/civicrm-core/blob/2af1ebd85dc162ff931de327634bef0516e3695b/CRM/Core/Form.php#L782

You may be able to address by adding a function to the form like in https://github.com/civicrm/civicrm-core/pull/30445/files

```
 /**
   * @return array
   */
  protected function getFieldsToExcludeFromPurification(): array {
    return ['// your field name here// ];
  }
```

**Comment by jofranz** [comment]
CC: @frank-systopia
This feels similar the problem we had with CkEditor a while ago

**Comment by jensschuppe** [comment]
> It maybe that setPurifiedDefaults is doing this https://github.com/civicrm/civicrm-core/blob/2af1ebd85dc162ff931de327634bef0516e3695b/CRM/Core/Form.php#L782
> 
> You may be able to address by adding a function to the form like in https://github.com/civicrm/civicrm-core/pull/30445/files
> 
> ```
>  /**
>    * @return array
>    */
>   protected function getFieldsToExcludeFromPurification(): array {
>     return ['// your field name here// ];
>   }
> ```

Thanks for the hint, @eileenmcnaughton! So this seems to potentially break all extension forms that store HTML in text fields …

**Comment by jofranz** [comment]
I can't seem to reproduce it with the current CiviCRM `5.74.4`.
I'm running both DonRec `2.2.8` and [this version](https://github.com/alietz/de.systopia.donrec/commit/9d2a68f150c0a0d158467df4429e51df48096b4d) of @alietz.
I've been using a clean install of DonRec with it's provided standard template. However, the code looks identical after saving. Tags like `<body>` are not stripped.

@romymy Could you verify if `2.2.8`  is working CiviCRM `5.74.4`?

**Comment by Peter-Reck** [comment]
Adding Andy

Von: jofranz ***@***.***>
Gesendet: Donnerstag, 20. Juni 2024 11:49
An: systopia/de.systopia.donrec ***@***.***>
Cc: Peter Reck ***@***.***>; Manual ***@***.***>
Betreff: Re: [systopia/de.systopia.donrec] Problems with PDF Template Code in Donrec Profiles since CiviCRM 5.72.0 (Issue #196)


I can't seem to reproduce it with the current CiviCRM 5.74.4.
I'm running both DonRec 2.2.8 and this version<https://github.com/alietz/de.systopia.donrec/commit/9d2a68f150c0a0d158467df4429e51df48096b4d> of @alietz<https://github.com/alietz>.
I've been using a clean install of DonRec with it's provided standard template. However, the code looks identical after saving. Tags like <body> are not stripped.

@romymy<https://github.com/romymy> Could you verify if 2.2.8 is working CiviCRM 5.74.4?

**Comment by jofranz** [comment]
I was wrong.
I can confirm:
- 2.2.8 is broken on `5.74.4` (and probably since roughly `5.72.1`)
- [this PR](https://github.com/alietz/de.systopia.donrec/commit/9d2a68f150c0a0d158467df4429e51df48096b4d) by @alietz fixes it running `5.74.4`.

@jensschuppe Would be great if you could merge it and tag a new version!

**Comment by jensschuppe** [comment]
Released with version `2.2.9`. Thanks all for helping to fix this!
